src/algo.rs

```diff
@@ -1117,200 +1284,382 @@ impl Forest {
+                    AlignSelf::FlexStart => {
+                        if constants.is_wrap_reverse {
+                            free_cross_space - constants.padding_border.cross_end(constants.dir)
+                        } else {
+                            constants.padding_border.cross_start(constants.dir)
                         }
-                        AlignSelf::FlexEnd => {
-                            if is_wrap_reverse {
-                                padding_border.cross_start(dir)
-                            } else {
-                                free_cross_space - padding_border.cross_end(dir)
-                            }
+                    }
+                    AlignSelf::FlexEnd => {
+                        if constants.is_wrap_reverse {
+                            constants.padding_border.cross_start(constants.dir)
+                        } else {
+                            free_cross_space - constants.padding_border.cross_end(constants.dir)
                         }
-                        AlignSelf::Center => free_cross_space / 2.0,
-                        AlignSelf::Baseline => free_cross_space / 2.0, // Treat as center for now until we have baseline support
-                        AlignSelf::Stretch => {
-                            if is_wrap_reverse {
-                                free_cross_space - padding_border.cross_end(dir)
-                            } else {
-                                padding_border.cross_start(dir)
-                            }
+                    }
+                    AlignSelf::Center => free_cross_space / 2.0,
+                    AlignSelf::Baseline => free_cross_space / 2.0, // Treat as center for now until we have baseline support
+                    AlignSelf::Stretch => {
+                        if constants.is_wrap_reverse {
+                            free_cross_space - constants.padding_border.cross_end(constants.dir)
+                        } else {
+                            constants.padding_border.cross_start(constants.dir)
                         }
                     }
-                };
+                }
+            };
 
-                self.nodes[child].layout = result::Layout {
-                    order: order as u32,
-                    size: result.size,
-                    location: Point {
-                        x: if is_row { offset_main } else { offset_cross },
-                        y: if is_column { offset_main } else { offset_cross },
-                    },
+            self.nodes[child].layout = result::Layout {
+                order: order as u32,
+                size: result.size,
+                location: Point {
+                    x: if constants.is_row { offset_main } else { offset_cross },
+                    y: if constants.is_column { offset_main } else { offset_cross },
+                },
+            };
+        }
+    }
+
+    fn compute_internal(
```

Comment:
This now has a much clearer high-level overview, which makes it easier to follow what is happening. Great work!